### ISSUE TITLE
Avoid ".[hash]" in html (for deving reasons) and choose if html-only

### ISSUE THREAD

**Issue by frederikhors** [posthtml/posthtml-hash#44]
@metonym two issues:

- I cannot use `v1.0.0` anymore with my https://github.com/sveltejs/template project because using `npm run dev` it searches for `bundle.[hash].css` (literally) and `bundle.[hash].js` (literally); I don't wanna use `posthtml-hash` during dev: can we disable this new "html" way of doing things?

- My request was not to use this declarative way of handle names in `index.html`. Instead my request was about finding a way to use [the default native Rollup feature](https://rollupjs.org/guide/en/#outputassetfilenames) of declaring `bundle.[hash].js` avoiding `posthtml-hash` for doing this on js files: I need only to replace js file name in `index.html`: something like:

```js
{
  js: "html-only",
  css: "both" //default
}
```

**Comment by metonym** [comment]
Hi @frederikhors 

- v1 definitely has breaking changes in the way `rollup.config.js` should be configured. Have you tried customizing the output file to be `bundle.[hash].js`?

```diff
// rollup.config.js
{
  output: {
    sourcemap: true,
    format: 'iife',
    name: 'app',
-   file: 'public/build/bundle.js'
+   file: 'public/build/bundle.[hash].js'
  }
}
```

I've updated the [svelte-rollup-template](https://github.com/metonym/svelte-rollup-template) to use v1.

- I think the bigger issue is that perhaps a rollup plugin is better suited to deal with hashing static assets for production. For your use case, are you able to revert back to [v0.3.0](https://github.com/posthtml/posthtml-hash/blob/master/docs/v0.3.0.md)?

---

I'm open to any PRs for v1.

**Comment by frederikhors** [comment]
Using [outputentryfilenames](https://rollupjs.org/guide/en/#outputentryfilenames) in `rollup.config.js` doesn't work because also in dev mode rollup uses hash in js name (I also tried your `svelte-rollup-template` but you're using `file: "build/bundle.[hash].js"`).

So I think we need a way to disable this new behavior in `v1.0.0`: it's annoying to have an alert for new packages versions in your dependencies.

Then we can think on how to do it.

**Comment by metonym** [comment]
@frederikhors The major version signifies that the API is not backwards compatible. I would suggest using v0.3.0 or forking this project.

**Comment by frederikhors** [comment]
I know that.

I think we could use the plugin both ways: with and without `[hash]` in html.

This I am asking.

**Comment by khuongduybui** [comment]
> Using [outputentryfilenames](https://rollupjs.org/guide/en/#outputentryfilenames) in `rollup.config.js` doesn't work because also in dev mode rollup uses hash in js name (I also tried your `svelte-rollup-template` but you're using `file: "build/bundle.[hash].js"`).

Can't you do it like this?

```javascript
const production = !process.env.ROLLUP_WATCH;
const hash = production ? '.[hash]' : '';

export default {
  input: 'src/main.js',
  output: {
    sourcemap: true,
    format: 'iife',
    name: 'app',
    file: `build/bundle${hash}.js`,
  },
  ...
```

**Comment by frederikhors** [comment]
Nope because in dev I need to find a file named `bundle.[hash].js` literally, in prod I need the hashed name. But using `output.entryFileNames` in `rollup.config.js` (which I am, not `file:`) changes `[hash]` in a real hash also in dev mode.

**Comment by metonym** [comment]
@frederikhors What does your `rollup.config.js` look like?

**Comment by frederikhors** [comment]
@metonym it's like this:

```js
import svelte from 'rollup-plugin-svelte'
import resolve from '@rollup/plugin-node-resolve'
import commonjs from '@rollup/plugin-commonjs'
import livereload from 'rollup-plugin-livereload'
import { terser } from 'rollup-plugin-terser'
import fs from 'fs'
import posthtml from 'posthtml'
import { hash } from 'posthtml-hash'
import htmlnano from 'htmlnano'

const production = !process.env.ROLLUP_WATCH

export default {
  input: 'src/main.js',
  output: {
    sourcemap: false,
    format: 'esm',
    dir: 'public',
    entryFileNames: 'bundle.js',
    chunkFileNames: `js/bundle${production ? '.[hash]' : ''}.chunk.js`
  },
  plugins: [
    svelte({
      dev: !production,
      css: css => {
        css.write('public/build/bundle.css')
      }
    }),

    resolve({
      browser: true,
      dedupe: ['svelte']
    }),

    commonjs(),

    !production && serve(),

    !production && livereload('public'),

    production && terser(),

    production && postbuild()
  ],
  watch: {
    clearScreen: false
  }
}

function serve () {
  let started = false
  return {
    writeBundle () {
      if (!started) {
        started = true

        require('child_process').spawn('npm', ['run', 'start', '--', '--dev'], {
          stdio: ['ignore', 'inherit', 'inherit'],
          shell: true
        })
      }
    }
  }
}

function postbuild () {
  return {
    async writeBundle () {
      const html = fs.readFileSync('public/index.html')

      await posthtml([
        hash({ path: 'public', hashLength: 10 }),
        htmlnano()
      ]).process(html).then(result => fs.writeFileSync('public/index.html', result.html))
    }
  }
}
```

**Comment by metonym** [comment]
@frederikhors PostHTML plugins should not be used as Rollup plugins.

If you want the entire build process in one file, you could use [Rollup's JavaScript API](https://rollupjs.org/guide/en/#rolluprollup):

```js
// custom-rollup.config.js
const rollup = require('rollup');

async function dev() {
  await rollup.watch({...});
}

async function build() {
  const bundle = await rollup.rollup({...});
  const { output } = await bundle.generate({...});
  await postbuild();
};
```

**Comment by frederikhors** [comment]
Oops! Thanks @metonym.

Anyway the problem I found is still there.

I can't use `entryFileNames: 'bundle.js'` with `v1.0.0`.

How to fix it?

I think we should allow the usage of posthtml-hash in `rollup.config.js` only, like in `v0.3.0`.

Please. 😢

**Comment by metonym** [comment]
@frederikhors As I understand it, you use Rollup's `output.chunkFileNames` for code splitting but you also want to hash the `bundle.css` emitted by svelte. What is the error emitted by Rollup for the code splitting part?

**Comment by frederikhors** [comment]
Oh, I guess I didn't explain myself from the first moment.

Using [your template](https://github.com/metonym/svelte-rollup-template) it works (although it is a bit "strange" for me to have files in development that are called literally: `bundle.[hash].js` and `bundle.[hash].css`).

Using your `index.html` file with my `rollup.config.js` doesn't work because it (both for dev and prod build) creates files with hash in name, e.g.: `bundle.f43f10c3.js`.

One solution is to use `posthtml-hash` in javascript way only, like in `v0.3.0`.

Is the problem more understandable now?

**Comment by metonym** [comment]
@frederikhors Thanks for clarifying. So you don't need to hash the JS files?

What does the output index.html file look like if you remove `posthtml-hash` from dev and prod?

**Comment by frederikhors** [comment]
I'm using the simple https://github.com/sveltejs/template.

I wanna use `posthtml-hash` as `postbuild` operation to make the simple `index.html` know the js and css files are hased (ONLY IN PROD!). In dev mode I'm fine using `bundle.js` and `bundle.css` (Rollup is also faster with no hashing process on every save!)

It works amazing well in `v0.3.0`.

My question in `v0.3.0` was about a feature to add in `posthtml-hash` to make optional the hashing of `js` or `css` file **and/or** the `REPLACE` in `index.html`.

If Rollup gives me an hashed .js file I only need to use `posthtml-hash` to replace its name in `index.html`.

It it clearer now?

**Comment by metonym** [comment]
@frederikhors Hmm... what's still unclear to me is why you would need to hash the JS file name after building if Rollup already does it for you.

But if v0.3.0 works, why not stay with v0.3.0?

**Comment by frederikhors** [comment]
> Why you would need to hash the JS file name after building if Rollup already does it for you.

Because if I use `entryFileNames: 'bundle.js'` it doesn't!

If I use `entryFileNames: 'bundle.[hash].js'` it does but I cannot use it in DEV mode because my `index.html` is literally searching for `bundle.[hash].js` (LITERALLY! no hashed name!).

> But if v0.3.0 works, why not stay with v0.3.0?

Because `v1.0.0` has this problem I'm trying to fix with your help and because every CI I use warns me about this update: it's very annoying especially if you have many projects.

```
Major Update Potentially breaking API changes. Use caution. posthtml-hash 0.3.0  ❯  1.0.0  https://github.com/metonym/posthtml-hash
```

**Comment by frederikhors** [comment]
However if this problem of mine (common to all who will use `entryFileNames` in Rollup) is a problem for you I will stop writing, I will close the issue and I will use another solution.

No problem.

I don't want to burden your day.  😄

**Comment by metonym** [comment]
@frederikhors 

> If I use entryFileNames: 'bundle.[hash].js' it does but I cannot use it in DEV mode because my index.html is literally searching for bundle.[hash].js (LITERALLY! no hashed name!).

FWIW, in the [template](https://github.com/metonym/svelte-rollup-template/blob/master/public/index.html), the index.html uses the raw `bundle.[hash].js` format. Then, the [Rollup config](https://github.com/metonym/svelte-rollup-template/blob/master/rollup.config.js#L17) outputs the JS bundle using the same format. This is designed so that the file hashing does not affect DEV. Only after building is the postbuild script executed.

To silence the CI warnings, have you tried pinning the dependency version?

**Comment by frederikhors** [comment]
> To silence the CI warnings, have you tried pinning the dependency version?

yep

> FWIW, in the template, the index.html uses the raw bundle.[hash].js format. Then, the Rollup config outputs the JS bundle using the same format. This is designed so that the file hashing does not affect DEV. Only after building is the postbuild script executed.

I know. `entryFileNames` is the problem.

**Comment by frederikhors** [comment]
Closing in favor of https://github.com/posthtml/posthtml-hash/issues/45.
